thisnotthat/map_cluster_labelling.py

```diff
@@ -1380,3 +1414,187 @@ def labels_for_display(self):
             [self.label_formatter(label) for label in label_layer]
             for label_layer in self.labels
         ]
+
+
+class SparseMetadataLabelLayers(object):
+    """Generate multiple layers of labelling for a map based on a dataframe of metadata associated to points. Multiple
+    layers of clusters are generated, with higher level layers having larger more general clusters. Each cluster is
+    then labelled by training a one versus the rest classifier to discern the cluster in terms of the associated
+    metadata. The feature importances can then be used to label a cluster with the most discerning features.
+
+    Parameters
+    ----------
+     source_vectors: Array of shape (n_samples, n_features)
+         The original high dimensional vector representation of the data
+
+     map_representation: Array of shape (n_samples, n_map_features)
+         The map representation of the data
+
+    sparse_metadata: spmatrix
+        A sparse matrix of metadata associated to the points of data / map representation. Usually this is associated
+        with metadata that has a high number of features, and any given sample only has non-zero values for a small
+        number of features. A prime example is a bag-of-words representation of a corpus of documents.
+
+    fetaure_name_dictionary: dict
+        A dictionary mapping column indices of the sparse matrix to feature names. For example, if the sparse matrix
+        were the output of sklearn's ``CountVectorizer`` the dict would be
+        ``{idx: word for word, idx in model.vocabulary_.items()}``.
+
+    vector_metric: str (optional, default = "cosine")
+        The metric to use on the source vector space.
+
+    cluster_map_representation: bool (optional, default = False)
+        Whether to directly cluster the map representation, or use UMAP to generate a representation for clustering
```

Comment:
```suggestion
If set to True, cluster your raw map representation.  If set to False apply UMAP first to generate a representation for clustering.
```
It's less eloquent and equivalent but a touch clearer (in my head at least) as I had to read this twice.